src/model/model.nit

```diff
@@ -1201,6 +1273,265 @@ abstract class MType
+# instead. Else, you may end up with undefined behaviors.
+class MIntersectionType
+	super MTypeSet[MType]
+
+	# Intersect the specified types.
+	init with_operands(mmodule: MModule, operands: MType...) do
+		init(mmodule, new HashSet[MType].from(operands))
+	end
+
+	redef fun ==(other)
+	do
+		# If we don’t redefine `==`, we would need to cache all intersections
+		# in the `Model` object (in order to make fixed-point algorithms work).
+		if super then return true
+		return other isa MIntersectionType and other.operands == operands
+	end
+
+	redef fun apply_to(operands, mmodule, anchor)
+	do
+		var i = operands.iterator
+		var result = i.item
+		i.next
+		for operand in i do
+			result = result.intersection(operand, mmodule, anchor)
+		end
+		return result
+	end
+
+	redef var as_notnull is lazy do
+		var new_operands = new Set[MType]
+		# De we need to add `Object` to the operands?
+		var add_object = true
+		for mtype in operands do
+			if mtype isa MClassType or mtype isa MNotNullType then
+				# Already not null.
+				return self
+			else
+				var constraint = mtype.undecorate
+				if constraint isa MNullType then
+					return model.bottom_type
+				else if constraint isa MClassType or
+						constraint isa MNotNullType then
+					add_object = false
+				end
+				new_operands.add constraint
+			end
+		end
+		if add_object then
+			new_operands.add mmodule.object_type
+		end
+		return apply_to(new_operands, mmodule)
+	end
+
+	# TODO
+	#redef var c_name is lazy do return…
+
+	redef fun can_be_null(mmodule, anchor)
+	do
+		for mtype in operands do
+			if not mtype.can_be_null(mmodule, anchor) then
+				return false
+			end
+		end
+		return true
+	end
+
+	redef fun collect_mclassdefs(mmodule)
+	do
+		var result = collect_mclassdefs_cache.get_or_null(mmodule)
+		if result == null then
+			result = new Set[MClassDef]
+			for mtype in operands do
+				result.add_all(mtype.collect_mclassdefs(mmodule))
+			end
+			collect_mclassdefs_cache[mmodule] = result
+		end
+		return result
+	end
+
+	private var collect_mclassdefs_cache = new Map[MModule, Set[MClassDef]]
+
+	redef fun collect_mclasses(mmodule)
+	do
+		var result = collect_mclasses_cache.get_or_null(mmodule)
+		if result == null then
+			result = new Set[MClass]
+			for mtype in operands do
+				result.add_all(mtype.collect_mclasses(mmodule))
+			end
+			collect_mclasses_cache[mmodule] = result
+		end
+		return result
+	end
+
+	private var collect_mclasses_cache = new Map[MModule, Set[MClass]]
+
+	redef fun collect_mtypes(mmodule)
+	do
+		var result = collect_mtypes_cache.get_or_null(mmodule)
+		if result == null then
+			result = new Set[MClassType]
+			for mtype in operands do
+				result.add_all(mtype.collect_mtypes(mmodule))
+			end
+			collect_mtypes_cache[mmodule] = result
+		end
+		return result
+	end
+
+	private var collect_mtypes_cache = new Map[MModule, Set[MClassType]]
+
+	redef var hash is lazy do return operands.hash
+
+	redef var undecorate is lazy do
```

Comment:
Haha! Now I remember why: it’s because the implementation will be totally different for unions when they will replace `NullabeType` (simply remove the `NullType` from the operands). I thought by keeping `undecorate` abstract for `MTypeSet`, at least I won’t forgot to redefine it for unions.